File: 0042-trapping-rain-water/0042-trapping-rain-water.js
const trap = (height) => {
  let storedWater = 0;

  let leftMax = 0;
  let rightMax = 0;

  let left = 0;
  let right = height.length - 1;

  while (left <= right) {
    if (height[left] < height[right]) {
      if (height[left] <= leftMax) {
        storedWater += leftMax - height[left];
      } else {
        leftMax = height[left];
      }
      
      left++;
    } else {
      if (height[right] <= rightMax) {
        storedWater += rightMax - height[right];
      } else {
        rightMax = height[right];
      }
      
      right--;
    }
  }

  return storedWater;
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        stored_water = 0 

        left_max = 0
        right_max = 0

        left = 0
        right = len(height) - 1

        while left < right:
            if height[left] < height[right]:
                if height[left] < left_max:
                    stored_water += left_max - height[left]
                else:
                    left_max = height[left]
                left += 1
            else:
                if height[right] < right_max:
                    stored_water += right_max - height[right]
                else:
                    right_max = height[right]
                right -= 1

        return stored_water
